### PR TITLE
Fix shellcheck failures on clean.sh and cache_go_dirs.sh

### PR DESCRIPTION
**Pull request by xichengliudui**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Ref #72956

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: hack/make-rules/clean.sh [hunk 1]
```diff
@@ -28,11 +28,11 @@ CLEAN_PATTERNS=(
   "test/e2e/generated/bindata.go"
 )
 
-for pattern in ${CLEAN_PATTERNS[@]}; do
-  for match in $(find "${KUBE_ROOT}" -iregex "^${KUBE_ROOT}/${pattern}$"); do
+for pattern in "${CLEAN_PATTERNS[@]}"; do
+  while IFS= read -r -d '' match; do
```
Comment (xichengliudui) [inline hack/make-rules/clean.sh]: @BenTheElder 
It might be wrong here. Can you help me?

Comment (xichengliudui) [reply hack/make-rules/clean.sh]:
```
In ./hack/make-rules/clean.sh line 32:
  for match in $(find "${KUBE_ROOT}" -iregex "^${KUBE_ROOT}/${pattern}$"); do
               ^-- SC2044: For loops over find output are fragile. Use find -exec or a while read loop.

For more information:
  https://www.shellcheck.net/wiki/SC2044 -- For loops over find output are fr...
```

**Review by BenTheElder — APPROVED** [review]
/lgtm
/approve

**Pull request merged**